### PR TITLE
feat(media): support audio and video tag with multi sources

### PR DESCRIPTION
**Pull request by MrMurdock11**
## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (refactoring and improving code)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Description
**Summary**: Support for audio with multiple sources and extends the existing video functionality to support multiple sources as well.
**Motivation**: I aim to enhance user experience by introducing a feature that allows content to be listened to. Recognizing the complexity of this feature, I've divided its development into several phases: embedding audio, creating custom audio controls, and integrating the feature into the front matter block. Currently, I've implemented audio embedding with support for multiple sources and have also upgraded the video functionality to accommodate multiple sources.

### CODE REVIEW

File: _includes/multi-sources.html
Comment (kungfux) [inline _includes/multi-sources.html:8]: This does not work for a path containing a dot and so may raise questions. Do you think it worse it to fix?

```diff
- {% assign extensions = supported_extensions | split: '|' %}
- {% assign path_struct = src | split: '.' %}
- {% assign base_path = path_struct.first %}
+ {% assign base_path = src | split: '.' | pop | join: '.' %}
+ {% assign original_media_type = src | split: '.' | last | downcase %}
```

File: _includes/embed/audio.html
Comment (kungfux) [inline _includes/embed/audio.html:1]: Do we really need to maintain `supported_audio_extensions`?
Can't we just pass any extension to `convert-media-type.html`, so it may return ready to use MIME type? Or is it done to prevent passing wrong extensions and misprints to `additional_types`?

**Review by kungfux — CHANGES_REQUESTED** [review]
@MrMurdock11  I'm looking forward to use this feature in my blog💪

File: _includes/embed/audio.html
Comment (MrMurdock11) [reply _includes/embed/audio.html:1]: Yes, It is. The `supported_(audio|video)_extnames` is needed in order to correctly specify file types and their media type to add the source tag. This way I am trying to avoid adding sources with the wrong data type for the audio or video tag.

File: _includes/embed/audio.html
Comment (MrMurdock11) [reply _includes/embed/audio.html:3]: I'm thinking of `extnames`. What do you think?
```suggestion
{% assign extnames = include.extnames | default: '' | strip | split: '|' %}
```

File: _includes/embed/audio.html
Comment (MrMurdock11) [reply _includes/embed/audio.html:1]: I decided to delete supported_(audio|video)_extnames.

File: _includes/convert-media-type.html
Comment (MrMurdock11) [reply _includes/convert-media-type.html:1]: Done

File: _includes/multi-sources.html
Comment (MrMurdock11) [reply _includes/multi-sources.html:1]: done

File: _includes/multi-sources.html
Comment (MrMurdock11) [reply _includes/multi-sources.html]: I rewrote the implementation and now the code supports even this path format `relative.path/to/my.audio/ai.audio.mp3` and will correctly split the path into the necessary parts

**Review by cotes2020 — COMMENTED** [review]
Thanks @MrMurdock11, I understand your motivation and am in favor of this new feature.

But this PR contains some unnecessary files. It's not a good practice to overuse `{% include %}`, it will slow down the Jekyll build.

**Comment by MrMurdock11** [comment]
Ok, I get it @cotes2020. I'll take a look at it and try to minimize the include calls.

**Comment by cotes2020** [comment]
BTW, please make good use of `git rebase` and `git force push` when updating your feature branch, as this will facilitate code review.

**Comment by MrMurdock11** [comment]
@cotes2020 I removed unnecessary files from `embed` directory and `{% include %}` tags. I decided to work with `_data` and move extensions and MIME types to allow us to dynamically reference and override extensions and MIME types as needed.

File: _includes/embed/video.html
Comment (kungfux) [inline _includes/embed/video.html:45]:
```suggestion
      <source src="{{ video_url }}" type="video/{{ media_item['mimeType'] }}">
```

File: _includes/embed/video.html
Comment (kungfux) [inline _includes/embed/video.html:47]:
```suggestion
      <source src="{{ video_url }}" type="video/{{ extname }}">
```

File: _includes/embed/video.html
Comment (kungfux) [inline _includes/embed/video.html:55]:
```suggestion
        <source src="{{ extra_src }}" type="video/{{ media_item['mimeType'] }}">
```

File: _includes/embed/video.html
Comment (kungfux) [inline _includes/embed/video.html:57]:
```suggestion
        <source src="{{ extra_src }}" type="video/{{ extra_extnames }}">
```

File: _posts/2019-08-08-write-a-new-post.md
Comment (kungfux) [inline _posts/2019-08-08-write-a-new-post.md:495]: I would suggest to make a change for consistency

```suggestion
## Audios
```

File: _posts/2019-08-08-write-a-new-post.md
Comment (kungfux) [inline _posts/2019-08-08-write-a-new-post.md:497]: I would suggest to make a change for consistency

```suggestion
### Audio File
```

File: _posts/2019-08-08-write-a-new-post.md
Comment (kungfux) [inline _posts/2019-08-08-write-a-new-post.md:507]:
```suggestion
#### Adding Additional Audio Formats
```

File: _posts/2019-08-08-write-a-new-post.md
Comment (kungfux) [inline _posts/2019-08-08-write-a-new-post.md:502]: I would review and update documentation for audio section to looks closer to the documentation for video. It makes it easier to read one after another when they look similar.

**Review by kungfux — CHANGES_REQUESTED** [review]
Having `media.yml` makes it super easy to override and extend to support more media types. I like it 🫶

File: _includes/embed/audio.html
Comment (kungfux) [reply _includes/embed/audio.html]: I personally prefer single word naming like `sources`, `extras`, or `types`.

File: _includes/embed/video.html
Comment (MrMurdock11) [reply _includes/embed/video.html:45]: thanks @kungfux

File: _includes/embed/video.html
Comment (MrMurdock11) [reply _includes/embed/video.html:47]: thanks @kungfux

File: _includes/embed/video.html
Comment (MrMurdock11) [reply _includes/embed/video.html:55]: thanks @kungfux

File: _includes/embed/audio.html
Comment (kungfux) [inline _includes/embed/audio.html:7]:
```suggestion
    {% include img-url.html src=src img_path=page.img_path %}
```

We have no dedicated alternative variable for video and audio at the moment like `img_path` for images. However, please consider to unify behavior between video and audio embedding: use `img_path` for audio, or remove for video. Would be great to have separate dedicated variable for video and audio later. Taking into account the warning in documentation, it may make sense to remove usage of `img_url` for video embedding.

File: _data/media.yml
Comment (kungfux) [inline _data/media.yml:8]: Nit: Consider adding more [common types](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types) right away

```suggestion
  mimeType: x-matroska
- extension: ogv
  mimeType: ogg
- extension: weba
  mimeType: webm
- extension: 3gp
  mimeType: 3gpp
- extension: 3g2
  mimeType: 3gpp2
- extension: mid
  mimeType: midi
```

**Review by kungfux — CHANGES_REQUESTED** [review]

**Review by kungfux — APPROVED** [review]
Thank you for your patience and commitment.

Please wait for @cotes2020 approval as he may have good insights.

File: _includes/embed/audio.html
Comment (cotes2020) [inline _includes/embed/audio.html:7]: I noticed that you use `img-url.html` in the embed audio/video, which may be confusing for future maintenance.

How about changing `img-url.html` to `media-url.html` ?
In that case we could also change `page.img_url` to `page.media_url`, `site.img_cdn` to `site.cdn`.

Comment (kungfux) [reply _includes/embed/audio.html:7]: It's fine to rename `img-url.html`, but the problem with `page.img_url` is more complicated and should be handled separately, I think.

Currently, `page.img_url` is not being passed to `img-url.html`, so users may specify a path to video/audio without the need to use full address starting from `https`.

I started using the video embed solution and found out that I cannot use `page.img_url` for video files. @MrMurdock11 mentioned this problem in the documentation here: PWA fails to cache chunked data. In general it might be a bad idea to cache all video/audio files. So I was thinking more in the direction of introducing separate variables for audio/video files. In my blog instance, I had to create custom embed code to achieve it, so I can store video files in a separate folder while not using CDN.

Comment (cotes2020) [reply _includes/embed/audio.html:7]: So do you think `site.img_cdn` can be used by video/audio?

Comment (kungfux) [reply _includes/embed/audio.html:7]: `site.img_cdn` allows us to point to a different base URL like `https://my-cdn.com` instead of the blog URL, say `https://my-blog.com`.  So from an embed perspective we are still able to specify a different folder e.g. `{% include embed/video.html src='/video/myvideo.mp4' %}` which will resolve to `https://my-cdn.com/video/myvideo.mp4`.

The problem occurs when we add `page.img_url` to this chain. For example, `https://my-cdn.com/images/video/myvideo.mp4`. Using `page.img_url` forces us to put video/audio files in the same directory as images.

I suppose this could still be excluded from the PWA cache if we host video files under `/images/video`, but it makes it impossible to archive if we don't use a subdirectory.

So to make it easier, it might make sense to introduce separate variables.

Another option might be to allow PWA to cache video/audio files by fixing the problem in the caching logic. I'm not sure at this point what the best practice is for PWA applications in such cases.

Comment (cotes2020) [reply _includes/embed/audio.html:7]: I get your point. So for now, we can keep `page.img_url` and only change the following filenames and variables:

- Change `_includes/img-url.html` to `_includes/media-url.html`
- Change `site.img_cdn` to `site.cdn`

Comment (MrMurdock11) [reply _includes/embed/audio.html:7]: I don't think these changes need to be implemented in this pull request, because these changes will cause fixes in a lot of files that have nothing to do with this feature. I suggest we close this pull request, and then I will create a new one with changes to `img_cdn` and `img-url`.

Comment (cotes2020) [reply _includes/embed/audio.html:7]: As you wish @MrMurdock11 :)

**Comment by kungfux** [comment]
@cotes2020  Don't get me wrong, I'm not pushing anyone, but I would like to ask who should merge this? Does @MrMurdock11 already have enough permissions? I ask in case you're all waiting for my action. 😊

**Comment by cotes2020** [comment]
@kungfux As per @MrMurdock11‘s idea he is going to close this PR and reopen a new one. I will be involved in another code review when that happens, as the changes are a bit much.

**Comment by MrMurdock11** [comment]
@cotes2020 Not really, I meant that we will finish this query and merge it, and then I will create a new one with `img-url` and `img_cdn`.

**Comment by kungfux** [comment]
I'm glad we cleared that up, because I had a different understanding of the answer too.

**Comment by cotes2020** [comment]
That was a misunderstanding on my part. 😂 

So hold on a bit longer and I'll review this PR in depth.

File: _includes/embed/audio.html
Comment (cotes2020) [inline _includes/embed/audio.html:35]: There is a lot of duplicated code in `audio.html` and `video.html`, and it may be necessary to add a common module to be referenced by both files (via `{% include %}`).

File: _includes/embed/video.html
Comment (cotes2020) [inline _includes/embed/video.html:40]:
```suggestion
    {% assign basename = filename | split: '.' | first %}
```

**Review by cotes2020 — CHANGES_REQUESTED** [review]

File: _includes/embed/audio.html
Comment (MrMurdock11) [reply _includes/embed/audio.html:35]: > Thanks @MrMurdock11, I understand your motivation and am in favor of this new feature.
> 
> But this PR contains some unnecessary files. It's not a good practice to overuse `{% include %}`, it will slow down the Jekyll build.

After this comment, I removed all the `{% include.... %}` and rebuilt the template for working without tag `include`

File: _includes/embed/video.html
Comment (MrMurdock11) [reply _includes/embed/video.html:40]: This option will only work if there are no more dots in the filename, for example: `audio.mp3`. But my implementation supports more complex paths, for example: `audio.ai.mp3`. In this case I get exactly the name with a suffix that the user can use.

File: _includes/embed/audio.html
Comment (cotes2020) [reply _includes/embed/audio.html:35]: Even if we don't use `{% include %}` here, the following redundant code can be avoided

```liquid 

    {% assign media_item = site.data.media | find: 'extension', extension %}
    {% if media_item %}
      <source src="{{ src }}" type="audio/{{ media_item['mimeType'] }}">
    {% else %}
      <source src="{{ src }}" type="audio/{{ extension }}">
    {% endif %}

    {% for extra_type in types %}
      {% assign filename = basename | append: '.' | append: extra_type %}
      {% assign extra_src = path | append: '/' | append: filename %}
      {% assign media_item = site.data.media | find: 'extension', extra_type %}
      {% if media_item %}
        <source src="{{ extra_src }}" type="audio/{{ media_item['mimeType'] }}">
      {% else %}
        <source src="{{ extra_src }}" type="audio/{{ extra_type }}">
      {% endif %}
    {% endfor %}
```

File: _includes/embed/video.html
Comment (cotes2020) [reply _includes/embed/video.html:40]: That makes sense

File: _includes/embed/audio.html
Comment (MrMurdock11) [reply _includes/embed/audio.html:35]: as you wish @cotes2020. I'll move it to sources template again.

Comment (cotes2020) [reply _includes/embed/audio.html:35]: Thanks for your patience @MrMurdock11, whether or not you end up moving to `sources`, the snippet I mentioned above should still be improved.

Comment (cotes2020) [reply _includes/embed/audio.html:35]: The way I figured it out is to add the `type` from the `src` to the `types`, then generate the HTML in a forloop.

Comment (MrMurdock11) [reply _includes/embed/audio.html:35]: Thank you for the feedback, @cotes2020! Ok, I get your point
```liquid
<video class="embed-video file" {{ poster }} {{ attributes }}>
    {% assign parts = video_url | split: '/' %}
    {% assign filename = parts | last %}
    {% assign extension = filename | split: '.' | last %}
    {% assign basename = filename | split: '.' | pop | join: '.' %}
    {% assign path = parts | pop | join: '/' %}

    {% assign types = types | push: extension %}

    {% for type in types %}
      {% assign filename = basename | append: '.' | append: type %}
      {% assign extra_src = path | append: '/' | append: filename %}
      {% assign media_item = site.data.media | find: 'extension', type %}
      {% if media_item %}
        <source src="{{ extra_src }}" type="video/{{ media_item['mime_type'] }}">
      {% else %}
        <source src="{{ extra_src }}" type="video/{{ type }}">
      {% endif %}
    {% endfor %}

    Your browser does not support the video tag. Here is a
    <a href="{{ video_url | strip }}">link to the video file</a> instead.
  </video>
```

**Comment by MrMurdock11** [comment]
@cotes2020 @kungfux Thank you for your feedback

**Review by cotes2020 — APPROVED** [review]
LGTM

File: _includes/embed/video.html
Comment (MrMurdock11) [reply _includes/embed/video.html]: Done

**Pull request merged**
